TIR test suite.

vext01 opened the pull request:
This change adds a rustc flag `--emit yk-tir` which causes ykrustc to
dump the TIR in a textual format and exit.

Then we add a test new test suite kind which allows us to write tests
against the textual TIR dumps. It works in much the same way as the
existing MirOpt test suite.

Note that this needs a ykpack change to go in first. Raising a draft PR for initial discussion.

What do you think of the general approach?

[I think `generate_tir()` can be tidied up some in light of this change, but I didn't want to do that in this PR].

vext01 commented:
(I don't know why the build is marked as failed)

vext01 commented:
(There shouldn't be a build yet)

ltratt commented:
Do you want to squash?

vext01 commented:
Why, I'm glad you asked, I most certainly do!

ltratt commented:
Go ahead!

vext01 commented:
splat.

ltratt commented:
bors try

vext01 commented:
bors delegate=ltratt

This should give you permission to do the `r+` later. We can add you to a list of admins who always have this permission.

vext01 commented:
that finished quicker than i expected.

ltratt commented:
It doesn't look like it's run the tests?

vext01 commented:
Yeah, that's expected.

I've configured buildbot to run `.buildbot.sh`. This means that we can version control the CI steps themselves, and we don't have to change the buildbot config to make changes to the steps.

Anyway, the contents of the script currently are:
```
./x.py clean  # We don't clone afresh to save time and bandwidth.
RUST_BACKTRACE=1 ./x.py build --config .buildbot.toml --stage 1 -i src/libstd
```

So indeed, no testing is yet done.

We can turn them on if you like. I don't know what the outcome will be!

I'm not sure why the build is still orange. I'll check the configuration.

vext01 commented:
bors try

vext01 commented:
bors ping

vext01 commented:
bors retry

vext01 commented:
bors ping

vext01 commented:
bors try

ltratt commented:
Please squash.

vext01 commented:
splat

ltratt commented:
bors r+

vext01 commented:
That last commit silences that particular error, but I'm going to do a full test run on bencher8 before we kick off bors again.

vext01 commented:
Sadly the tests fail in the same way as in #10:

```
...
thread '[run-pass] run-pass/backtrace-debuginfo.rs' panicked at 'explicit panic', tools/compiletest/src/runtest.rs:3277:9
note: Run with `RUST_BACKTRACE=1` for a backtrace.
                                                                                 
---- [run-pass] run-pass/backtrace.rs stdout ----
                                                                                   
error: test compilation failed although it shouldn't!
status: signal: 6                                                                                                                                                                     
command: "/home/vext01/research/ykrustc_pv/build/x86_64-unknown-linux-gnu/stage1/bin/rustc" "/home/vext01/research/ykrustc_pv/src/test/run-pass/backtrace.rs" "--target=x86_64-unknow$
-linux-gnu" "--error-format" "json" "-Zui-testing" "-C" "prefer-dynamic" "-o" "/home/vext01/research/ykrustc_pv/build/x86_64-unknown-linux-gnu/test/run-pass/backtrace/a" "-Crpath" "$
O" "-Zunstable-options" "-Lnative=/home/vext01/research/ykrustc_pv/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "-g" "-L" "/home/vext01/research/ykrustc_pv/build/x86_64-$
nknown-linux-gnu/test/run-pass/backtrace/auxiliary"                                                                                                                                   
stdout:                                                                                                                                                                               
------------------------------------------                                                                                                                                            
                                                                                                                                                                                      
------------------------------------------                                                                                                              
stderr:                                   
------------------------------------------                                                                                                                                            
rustc: /home/vext01/research/hsaingkai-llvm/include/llvm/CodeGen/SlotIndexes.h:418: llvm::SlotIndex llvm::SlotIndexes::getInstructionIndex(const llvm::MachineInstr&) const: Assertio$
 `itr != mi2iMap.end() && "Instruction not found in maps."' failed.                                                                                                                   
                                                                                                                                                                                      
------------------------------------------                                                                              

thread '[run-pass] run-pass/backtrace.rs' panicked at 'explicit panic', tools/compiletest/src/runtest.rs:3277:9

---- [run-pass] run-pass/issues/issue-26484.rs stdout ----

error: test compilation failed although it shouldn't!
status: signal: 6
command: "/home/vext01/research/ykrustc_pv/build/x86_64-unknown-linux-gnu/stage1/bin/rustc" "/home/vext01/research/ykrustc_pv/src/test/run-pass/issues/issue-26484.rs" "--target=x86_6
4-unknown-linux-gnu" "--error-format" "json" "-Zui-testing" "-C" "prefer-dynamic" "-o" "/home/vext01/research/ykrustc_pv/build/x86_64-unknown-linux-gnu/test/run-pass/issues/issue-264
84/a" "-Crpath" "-O" "-Zunstable-options" "-Lnative=/home/vext01/research/ykrustc_pv/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "-g" "-L" "/home/vext01/research/ykrustc
_pv/build/x86_64-unknown-linux-gnu/test/run-pass/issues/issue-26484/auxiliary"
stdout:
------------------------------------------

------------------------------------------
stderr:
------------------------------------------
rustc: /home/vext01/research/hsaingkai-llvm/include/llvm/CodeGen/SlotIndexes.h:418: llvm::SlotIndex llvm::SlotIndexes::getInstructionIndex(const llvm::MachineInstr&) const: Assertion
 `itr != mi2iMap.end() && "Instruction not found in maps."' failed.

------------------------------------------
```

At this stage, I have no idea what the error message is saying, but t's likely that my debug labels have broken some compiler assumptions.

Obviously I need to look into that, but what do you want to do with this PR in the meantime?

Cheers

ltratt commented:
Do we have to turn debuglabels on to hit this error?

vext01 commented:
In the `config.toml`, I had `debuginfo = false`. The test suite is probably building some programs in debug mode.

ltratt commented:
Ah, I see. I'm not sure. In one sense I wouldn't mind merging this, but there again, I feel like we shouldn't in general be merging something that breaks tests (at least until we understand for sure why they're failing). Do we think we can fix this test failure, or is it something we'll have to live with for a while?

vext01 commented:
> I feel like we shouldn't in general be merging something that breaks tests

To be clear, the code that broke the tests was already merged. We've just exposed the failure by running the tests.

> Do we think we can fix this test failure, or is it something we'll have to live with for a while?

Hard to say I'm afraid.

It frustrates me slightly that I'll be chasing LLVM again, when we know that in the long term we want to ditch it.

I suggest I investigate this today and see what I can come up with and report back.

In the meantime, it's up to you about this PR. We could:
 * Merge it turning off tests.
 * Leave it open in the hope that I can fix this soon.

I'm pretty torn myself.

ltratt commented:
> To be clear, the code that broke the tests was already merged. We've just exposed the failure by running the tests.

Point taken. Let's merge this PR then investigate separately whether a) we're going to fix it or b) live with it until we can wean ourselves off LLVM.

[I assume that we thus need to disable bors from running tests?]

vext01 commented:
> [I assume that we thus need to disable bors from running tests?]

Yep.

By the way, were you happy with the concept and name of the `.buildbot.sh` script?

ltratt commented:
Happy with both.

If you push a commit that disables tests, I expect we can get this done soon.

vext01 commented:
That should do it. Can I squash it in?

(I'll leave the style changes separate as they are nothing to do with CI per-se)

ltratt commented:
Yep!

vext01 commented:
Squashed. You should be able to get bors to auto-merge this now.

ltratt commented:
bors r+

vext01 committed:
TIR test suite.

This change adds a rustc flag `--emit yk-tir` which causes ykrustc to
dump the TIR in a textual format and exit.

Then we add a test new test suite kind which allows us to write tests
against the textual TIR dumps. It works in much the same way as the
existing MirOpt test suite. In fact, we re-use a lot of that code.

ltratt commented on src/librustc_interface/passes.rs:
```diff
@@ -1040,11 +1040,18 @@ pub fn start_codegen<'tcx>(
     // Output Yorick debug sections into binary targets.
     if tcx.sess.crate_types.borrow().contains(&config::CrateType::Executable) {
+        // Are we dumping the textual version of the TIR at the same time?
+        let tir_dump_fn = if tcx.sess.opts.output_types.contains_key(&OutputType::YkTir) {
```
`let tir_dump_fn = tcsx.sess...contains_key(...).map(|_| Some(outputs.path(OutputType::YkTir)));` or similar should work.

ltratt commented on src/librustc_yk_sections/mir_cfg.rs:
```diff
@@ -127,6 +128,10 @@ pub fn generate_tir<'a, 'tcx, 'gcx>(
     sorted_def_ids.sort();
 
+    // Only if the user requested, open a file for the textual TIR dump.
+    let mut dump_file: Option<File> = dump_filename.map_or_else::<Result<Option<File>, io::Error>, _, _>(
```
Are the type annotations needed? They make this very hard to read.

vext01 replied:
Hah, ironically I added them because I found it hard to read.

I'm not proud of this line. I think I should unroll it into something simpler.

ltratt replied:
I'd just drop the annotations myself and then I think it'll be fine.

ltratt commented on src/tools/compiletest/src/runtest.rs:
```diff
@@ -3258,6 +3266,45 @@ impl<'test> TestCx<'test> {
+            self.fatal_proc_rec("compilation failed!", &proc_res);
+        }
+
+        if !proc_res.status.success() {
```
This is the same condition as the `if` above which seems odd.

vext01 replied:
Oops. This is a copy/paste mistake. Nice catch.

ltratt commented on src/tools/compiletest/src/runtest.rs:
```diff
@@ -3258,6 +3266,45 @@ impl<'test> TestCx<'test> {
+            let mut test_lines = vec![ExpectedLine::Elision];
+            for l in test_text.lines() {
+                if l.is_empty() {
+                    // ignore
```
It might be easier to put `continue` in here. Or maybe `for l in test_text.lines().filter(|x| !x.is_empty())` is even clearer?

vext01 replied:
Just following the precedent set in the `MirOpt` suite. This function was a copy-paste-modify kind of affair.

vext01 replied:
Here's the code this was based on in rustc:
https://github.com/rust-lang/rust/blob/237bf3244fffef501cf37d4bda00e1fce3fcfb46/src/tools/compiletest/src/runtest.rs#L2944-L2946

Might as well follow upstream.

ltratt commented on src/tools/compiletest/src/runtest.rs:
```diff
@@ -3258,6 +3266,45 @@ impl<'test> TestCx<'test> {
+            for l in test_text.lines() {
+                if l.is_empty() {
+                    // ignore
+                } else if l.starts_with("//") && l.split_at("//".len()).1.trim() == "..." {
```
This feels like it would be clearer (and faster) as something like:

```rust
if l.starts_with("//") {
   if l["".len()..].trim() == "..." { ... }
   else { ... }
}
```

vext01 replied:
As before, let's follow upstream's example.

ltratt commented:
The test looks pretty much as I expected/hoped, so this looks like it's going in the right direction to me.!

vext01 commented:
Great!

Do you think I should take the time here to tidy up `generate_mir`, which could accept an enum for the output type (binary or text)?

It's not essential.

ltratt commented:
> Do you think I should take the time here to tidy up generate_mir, which could accept an enum for the output type (binary or text)?

Dunno.

vext01 commented:
OK, let me have a play and see if I can get some tidying done fairly quickish.

vext01 commented:
> see if I can get some tidying done

The last commit makes the interface and internals of `generate_tir()` more palatable. Only compile tested at the moment. Testing now.

This was surprisingly hard to get through the borrow checker!

ltratt commented on src/librustc_yk_sections/mir_cfg.rs:
```diff
@@ -112,26 +121,59 @@ impl<'a, 'tcx, 'gcx> ConvCx<'a, 'tcx, 'gcx> {
+            // In this case the file at `tir_path` is a raw binary file which we use to make an
+            // object file for linkage.
+            let obj = YkExtraLinkObject::new(&tir_path, SECTION_NAME);
+            fs::remove_file(tir_path)?; // Now we have our object, we can remove the temp file.
```
We might want to suppress error if the temporary file can't be removed (it's not the end of the world)?

vext01 replied:
Up to you. It's super unlikely to happen.

ltratt commented on src/librustc_yk_sections/mir_cfg.rs:
```diff
@@ -112,26 +121,59 @@ impl<'a, 'tcx, 'gcx> ConvCx<'a, 'tcx, 'gcx> {
+        },
+    };
+
+    let mut enc = match default_file {
```
`let mut enc = default_file.map(|ref mut f| Some(...));` or similar?

vext01 replied:
Ah, yes.

vext01 replied:
Hrm.

Seems the closure upsets the checker:
```
167 |     let mut enc = default_file.map(|ref mut f| ykpack::Encoder::from(f));
    |                                     ^^^^^^^^^                         - temporary value dropped here while still borrowed
    |                                     |
    |                                     temporary value does not live long enough
```

Any ideas?

ltratt replied:
Although I doubt it works, does `|mut f|` fix things?

vext01 replied:
Funnily enough, if I remove the `ref` it asks me to also remove `mut`, so the closest we would get is:

```
169 |     let mut enc = default_file.map(|f| ykpack::Encoder::from(&mut f));
    |                                                                   ^- `f` dropped here while still borrowed
    |                                                                   |
    |                                                                   borrowed value does not live long enough
...
206 | }
    | - borrowed value needs to live until here
```

ltratt replied:
OK, let's give up ;)

vext01 commented:
One outstanding comment for you here:
https://github.com/softdevteam/ykrustc/pull/11#discussion_r269956205

Besides that, I think I've covered your other comments.

Note that the companion PR for ykpack needs to go in first:
https://github.com/softdevteam/ykpack/pull/7

vext01 commented:
[I'd like to add more/better tests, but we'd need to implement more of the TIR conversions to be able to make sense of TIR dumps -- Currently there are too many 'unimplemented's to know where in the code we are].

ltratt commented:
I think we're ready to squash?

vext01 commented:
Not just yet. One moment.

vext01 commented:
In addition to that last commit, I propose (as a separate commit) renaming `mir_cfg.rs` to `emit_tir.rs` or something similar.

Thoughts?

ltratt commented:
Fine with me.

vext01 committed:
Rename mir_cfg.rs to emit_tir.rs.

The old name was a throwback to when we though we would simply serialise
MIR, instead of designing our own tracing byte-code.

vext01 commented:
I've renamed the file and shut the tidy checker up.

OK for me to squash?

ltratt commented:
Please squash.

vext01 commented:
splat.

ltratt commented:
bors r+
